crates/dojo-defi/src/constant_product_market/components.cairo

```diff
@@ -40,6 +79,189 @@ impl MarketImpl of MarketTrait {
+            (amount, quantity)
+        } else {
+            // Given the amount, get optimal quantity to add to the market
+            let quantity_optimal = self.quote_quantity(amount);
+            if quantity_optimal <= quantity {
+                // Add the given amount and optimal quantity to the market
+                (amount, quantity_optimal)
+            } else {
+                let amount_optimal = self.quote_amount(quantity);
+                // Ensure that the optimal amount is less than or equal to the given amount
+                assert(amount_optimal <= amount, 'insufficient amount');
+                (amount_optimal, quantity)
+            }
+        }
+    }
+
+    // Add liquidity to the market, mints shares for the given amount of liquidity provided
+    //
+    // Arguments:
+    //
+    // amount: The amount of cash to add to the market
+    // quantity: The quantity of items to add to the market
+    //
+    // Returns:
+    //
+    // (amount, quantity, shares): The amount of cash and quantity of items added to the market and the shares minted
+    fn add_liquidity(self: @Market, amount: u128, quantity: usize) -> (u128, usize, FixedType) {
+        // Compute the amount and quantity to add to the market
+        let (amount, quantity) = self.add_liquidity_inner(amount, quantity);
+        // Mint shares for the given amount of liquidity provided
+        let shares = self.mint_shares(amount, quantity);
+        (amount, quantity, shares)
+    }
+
+    // Mint shares for the given amount of liquidity provided
+    fn mint_shares(self: @Market, amount: u128, quantity: usize) -> FixedType {
+        // If there is no liquidity, then mint total shares
+        if !self.has_liquidity() {
+            (Fixed::new_unscaled(amount, false)
+                * Fixed::new_unscaled(
+                    quantity.into().try_into().unwrap() * SCALING_FACTOR, false
+                )).sqrt()
+        } else {
+            // Convert amount to fixed point
+            let amount = Fixed::new_unscaled(amount, false);
+
+            // Get normalized reserve cash amount and item quantity
+            let (reserve_amount, _) = self.get_reserves();
+
+            // Convert reserve amount to fixed point
+            let reserve_amount = Fixed::new_unscaled(reserve_amount, false);
+
+            // Get total liquidity
+            let liquidity = self.liquidity();
+
+            // Compute the amount of shares to mint
+            // S = dx * L/X = dy * L/Y
+            (amount * liquidity) / reserve_amount
+        }
+    }
+
+    // Remove liquidity from the market, return the corresponding amount and quantity payout
+    //
+    // Arguments:
+    //
+    // shares: The amount of liquidity shares to remove from the market
+    //
+    // Returns:
+    //
+    // (amount, quantity): The amount of cash and quantity of items removed from the market
+    fn remove_liquidity(self: @Market, shares: FixedType) -> (u128, usize) {
```

Comment:
This implementation follows the one from [UniswapV2Pair](https://etherscan.io/address/0xb4e16d0168e52d35cacd2c6185b44281ec28c9dc). Cash amount and item quantity payout follows the formula `dx = S * X / L`, `dy = S * Y / L`, respectively and ensures ratio is intact when liquidity is removed